File: src/main/java/edu/colorado/plv/droidstar/experiments/lp/SQLiteOpenHelperLP.java
package edu.colorado.plv.droidstar.experiments.lp;

import java.io.File;
import java.util.List;
import java.util.ArrayList;
import java.util.Queue;
import java.util.ArrayDeque;

import android.os.Handler.Callback;
import android.content.Context;
import android.database.sqlite.SQLiteOpenHelper;
import android.database.sqlite.SQLiteDatabase;

import edu.colorado.plv.droidstar.LearningPurpose;
import static edu.colorado.plv.droidstar.Static.*;

public class SQLiteOpenHelperLP extends LearningPurpose {
    protected MyDBHelper helper;
    protected File testDB;

    public String shortName() {return "SQLiteOpenHelper";}

    protected File dbfile(Context c) {
        return new File(c.getExternalFilesDir(null), "testDB.sqlite3");
    }

    // INPUTS
    public static String OPENNEW = "openNew";
    public static String OPENHV = "openHV"; // higher ver
    public static String OPENLV = "openLV"; // lower ver
    public static String OPENSV = "openSV"; // same ver
    public static String CLOSE = "close";

    public static int THISV = 5;
    public static int HV = 6;
    public static int LV = 4;

    public List<String> uniqueInputSet() {
        List<String> is = new ArrayList();
        is.add(OPENNEW);
        is.add(OPENHV);
        is.add(OPENLV);
        is.add(OPENSV);
        is.add(CLOSE);

        return is;
    }

    // All queries must start with an OPEN input, and then can't have
    // any more OPEN inputs.  This is a very limited experiment...
    public boolean validQuery(Queue<String> q) {
        Queue<String> query = new ArrayDeque(q);

        String first = query.poll();

        if (first == null) return true;

        if (first.equals(OPENHV)
            || first.equals(OPENNEW)
            || first.equals(OPENLV)
            || first.equals(OPENSV)) {

            for (String input : query) {
                if (input.equals(OPENHV)
                    || input.equals(OPENNEW)
                    || input.equals(OPENLV)
                    || input.equals(OPENSV)) {
                    return false;
                }
            }
            return super.validQuery(query);
        }
        
        return false;
        
    }

    // OUTPUTS
    public static String CONFIGURED = "confd";
    public static String CREATED = "created";
    public static String OPENED = "opened";
    public static String UPGRADED = "upgrd";

    public SQLiteOpenHelperLP(Context c) {
        super(c);
        this.testDB = dbfile(c);
    }

    public boolean isError(String output) {return false;}

    public int betaTimeout() {
        return 500;
    }

    public void initdb(int v) {
        SQLiteDatabase db = SQLiteDatabase.openOrCreateDatabase(testDB, null);
        db.setVersion(v);
        db.close();
    }

    public void openinput(int v) {
        initdb(v);
        helper.getWritableDatabase();
    }

    public void giveInput(String input, int altKey) throws Exception {
        if (input.equals(OPENSV)) {
            openinput(THISV);
        } else if (input.equals(OPENNEW)) {
            helper.getWritableDatabase();
        } else if (input.equals(OPENHV)) {
            openinput(HV);
        } else if (input.equals(OPENLV)) {
            openinput(LV);
        } else if (input.equals(CLOSE)) {
            helper.close();
        }
    }

    protected String resetActions(Context context, Callback callback) {
        if (helper != null) helper.close();
        testDB.delete();
        helper = new MyDBHelper(context, testDB.getAbsolutePath(), null, 5);
        return null;
    }

    protected class MyDBHelper extends SQLiteOpenHelper {
        public MyDBHelper(Context c, String n, SQLiteDatabase.CursorFactory f, int v) {
            super(c,n,f,v);
        }

        public void onConfigure(SQLiteDatabase db) {respond(CONFIGURED);}
        public void onCreate(SQLiteDatabase db) {respond(CREATED);}
        public void onOpen(SQLiteDatabase db) {respond(OPENED);}
        public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
            respond(UPGRADED);
        }
    }
}

File: driver-app/src/main/scala/edu/colorado/plv/droidstar/experiments/lp/SQLiteOpenHelperLP.scala
package edu.colorado.plv.droidstar
package experiments.lp

import java.io.File;
import java.util.ArrayList;
import java.util.Queue;
import java.util.ArrayDeque;

import android.os.Handler.Callback;
import android.content.Context;
import android.database.sqlite.SQLiteOpenHelper;
import android.database.sqlite.SQLiteDatabase;


import scala.collection.JavaConverters._

class SQLiteOpenHelperLP(c: Context) extends LearningPurpose(c) {

  def dbfile(c: Context): File =
    new File(c.getExternalFilesDir(null), "testDB.sqlite3")

  var helper: MyDBHelper = null
  var testDB: File = dbfile(c)

  def initdb(v: Int): Unit = {
    var db: SQLiteDatabase = SQLiteDatabase.openOrCreateDatabase(testDB, null)
    db.setVersion(v)
    db.close()
  }

  // INPUTS
  val OPENNEW = "openNew"
  val OPENHV = "openHV" // higher ver
  val OPENLV = "openLV" // lower ver
  val OPENSV = "openSV" // same ver
  val CLOSE = "close"

  val THISV: Int = 5
  val HV: Int = 6
  val LV: Int = 4

  // OUTPUTS
  val CONFIGURED = "confd"
  val CREATED = "created"
  val OPENED = "opened"
  val UPGRADED = "upgrd"

  override def betaTimeout(): Int = 500

  def openinput(v: Int): Unit = {
    initdb(v)
    helper.getWritableDatabase()
  }

  @throws(classOf[Exception])
  override def giveInput(i: String, k: Int): Unit = i match {
    case `OPENNEW` => helper.getWritableDatabase()
    case `OPENSV` => openinput(THISV)
    case `OPENHV` => openinput(HV)
    case `OPENLV` => openinput(LV)
    case `CLOSE` => helper.close()
    case _ => {
      logl("Unknown command to SQLiteOpenHelper")
      throw new IllegalArgumentException("Unknown command to AsyncTask")
    }
  }

  // All queries must start with an OPEN input, and then can't have
  // any more OPEN inputs.  This is a very limited experiment...
  override def validQuery(q: Queue[String]): Boolean = {
    var query: Queue[String] = new ArrayDeque(q)
    val first: String = query.poll()

    if (first == null) {
      return true
    }
    if (first.equals(OPENHV)
        || first.equals(OPENNEW)
        || first.equals(OPENLV)
        || first.equals(OPENSV)) {

        for (input <- query.asScala) {
            if (input.equals(OPENHV)
                || input.equals(OPENNEW)
                || input.equals(OPENLV)
                || input.equals(OPENSV)) {
                return false
            }
        }
        return super.validQuery(query)
    }
    return false
  }

  override def isError(o: String): Boolean = false

  override def resetActions(c: Context, b: Callback): String = {
    if (helper != null) {
      helper.close()
    }
    testDB.delete()
    helper = new MyDBHelper(c, testDB.getAbsolutePath(), null, 5)
    null
  }

  class MyDBHelper(
    c: Context,
    n: String,
    f: SQLiteDatabase.CursorFactory,
    v: Int
  ) extends SQLiteOpenHelper(c,n,f,v) {
    override def onConfigure(db: SQLiteDatabase): Unit =
      respond(CONFIGURED)
    override def onCreate(db: SQLiteDatabase): Unit =
      respond(CREATED)
    override def onOpen(db: SQLiteDatabase): Unit =
      respond(OPENED)
    override def onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int): Unit =
      respond(UPGRADED)
  }

  override def shortName(): String = "SQLiteOpenHelper"

  override def uniqueInputSet(): java.util.List[String] =
    List(OPENNEW, OPENHV, OPENLV, OPENSV, CLOSE).asJava

}
